icu4c/source/common/uloc.cpp

```diff
@@ -1588,77 +1553,85 @@ _canonicalize(const char* localeID,
 
     origLocaleID=tmpLocaleID;
 
-    /* if we are doing a full canonicalization, then put results in
-       localeBuffer, if necessary; otherwise send them to result. */
-    if (/*OPTION_SET(options, _ULOC_CANONICALIZE) &&*/
-        (result == NULL || resultCapacity < (int32_t)sizeof(localeBuffer))) {
-        name = localeBuffer;
-        nameCapacity = (int32_t)sizeof(localeBuffer);
-    } else {
-        name = result;
-        nameCapacity = resultCapacity;
+    LocalMemory<char> localeBuffer;
+    LocalPointer<CheckedArrayByteSink> localSink;
+    internal::MemoryByteSink* sinkp = &sink;
```

Comment:
AFAICT this is always a CheckedArrayByteSink, so please use that, and again we should not need the new methods. We may need to read from the localeBuffer instead of from sinkp->data().

If we do need different sink types, then I would start with a variable that tells us where the contents is -- I am reluctant to add API to the ByteSink classes.